Tutorial.playground/Contents.swift

```diff
@@ -0,0 +1,662 @@
+//: documentation for `yield` says that you can also use `yield` within a
+//: `Pipe`, too, because of this polymorphism:
+//:
+//     func yield<UO, UI, DO>(@autoclosure(escaping) value: () -> DO) -> Proxy<UO, UI, (), DO, ()>
+//:
+//: Use simpler types like these to guide you until you understand the fully
+//: general type.
+
+//: `for_` loops are the simplest way to consume a `Producer` like `stdinByLine`.
+//: `for_` has the following type:
+//:
+//                        +-- Producer           +-- The body of the   +-- Result
+//                        |   to loop            |   loop              |
+//                        v   over               v                     v
+//                        --------------         ---------------       ---------
+//     func for_<A, R>(p: Producer<A, R>.T, _ f: A -> Effect<()>.T) -> Effect<R>.T
+//
+//: `for_(producer, body)` loops over `producer`, substituting each `yield` in
+//: `producer` with `body`.
+//:
+//: You can also deduce that behavior purely from the type signature:
+//:
+//: * The body of the loop takes exactly one argument of type `A`, which is
+//: the same as the output type of the `Producer`.  Therefore, the body of the
+//: loop must get its input from that `Producer` and nowhere else.
+//:
+//: * The return value of the input `Producer` matches the return value of the
+//: result, therefore `for_` must loop over the entire `Producer` and not skip
+//: anything.
+//:
+//: The above type signature is not the true type of `for_`, which is actually
+//: more general.  Think of the above type signature as saying: "If the first
+//: argument of `for_` is a `Producer` and the second argument returns an
+//: `Effect`, then the final result must be an `Effect`."
+//:
+//: Jump to the definition of (⌘+Click) `for_` to navigate to its documentation.
+//: There you will see the fully general type and underneath you will see equivalent 
+//: simpler types.  One of these says that if the body of the loop is a `Producer`, 
+//: then the result is a `Producer`, too:
+//:
+//     func for_<UO, UI, DI, DO, NI, NO, FR>(p: Proxy<UO, UI, DI, DO, FR>, _ f: DO -> Proxy<UO, UI, NI, NO, DI>) -> Proxy<UO, UI, NI, NO, FR>
+//:
+//: The first type signature we showed for `for_` was a special case of this
+//: slightly more general signature because a `Producer` that never `yield`s is
+//: also an `Effect`:
+//:
+//     `X` is an "uninhabited" type.  Practically, this means all attempts to
+//     construct an `X` will fail catastrophically.
+//     struct X { //... }
+//
+//: This is why `for` permits two different type signatures.  The first type
+//: signature is just a special case of the second one:
+//
+//     func for_<A, B, R>(p : Producer<A, R>, f : (A -> Producer<B, ()>) -> Producer<B, R> 
+//
+//     Specialize `B` to `X`
+//     func for_<A, R>(p : Producer<A, R>, f : (A -> Producer<X, ()>) -> Producer<X, R>
+//
+//     Producer<X, ?> == Effect
```

Comment:
What to do about this question mark?  Maybe

`Producer<X, Result> == Effect<Result>`